### DISCUSSION TITLE
Charge and dipole moment prediction

### DISCUSSION THREAD

**Discussion by 24ssuman** [mir-group/allegro#89] [General]
Hello devs,

I was just wondering, are there any ways to provide the atomic charges/ Wannier centroids in the training sets and train the Allegro with that to predict the charges and dipole moments from the simulations?

**Comment by cw-tan** [comment]
Hi @24ssuman, 

So there are two aspects to this.

1. For the case of data processing - yes, it's all set up and ready to go. You'd need to use the `extxyz` format if your dataset contains variable numbers of atoms. It would look something like
```
6
Properties=species:S:1:pos:R:3:dipole:R:3:forces:R:3 total_energy=-189.653357372 pbc="F F F"
C       -0.96304567      -0.72065914       0.02419782       0.00173113       0.00888203      -0.00512622       0.01852917       0.09030379      -0.00850674
C       -0.95342971       0.71275060       0.00811935      -0.00039242      -0.00271234      -0.01159632       0.01838543      -0.09894311       0.00389510
C        0.37129433       0.01484782      -0.00243504      -0.00304444      -0.00518352      -0.03681482       0.04654283      -0.00870709       0.00658218
O        1.57990998      -0.00410760      -0.00328219       0.00255312       0.00067816       0.01090803      -0.06682394       0.00398918      -0.00224618
H       -1.63063067      -1.52343335      -0.19415000       0.00091514      -0.00237370       0.02141257      -0.00430894      -0.01092267       0.00284952
H       -1.62806353       1.50604836      -0.11169229      -0.00176253       0.00070937       0.02121676      -0.01232454       0.02427991      -0.00257388
6
Properties=species:S:1:pos:R:3:dipole:R:3:forces:R:3 total_energy=-189.498015624 pbc="F F F"
C       -0.96304567      -0.72065914       0.02419782       0.00173113       0.00888203      -0.00512622       0.01261149       0.02284332       0.01899061
C       -0.95342971       0.71275060       0.00811935      -0.00039242      -0.00271234      -0.01159632       0.04223238      -0.04508272      -0.00158595
C        0.37129433       0.01484782      -0.00243504      -0.00304444      -0.00518352      -0.03681482      -0.09915466       0.00895942      -0.00673180
O        1.57990998      -0.00410760      -0.00328219       0.00255312       0.00067816       0.01090803       0.02688150      -0.00033351      -0.00014549
H       -1.63063067      -1.52343335      -0.19415000       0.00091514      -0.00237370       0.02141257       0.01453133      -0.01544667      -0.01012869
H       -1.62806353       1.50604836      -0.11169229      -0.00176253       0.00070937       0.02121676       0.00289795       0.02906016      -0.00039869
6
```
You can add in `charge` in an analogous way (since it's a per atom scalar). You'd probably also want to define the custom keys used (see [here](https://github.com/mir-group/nequip/blob/main/nequip/data/_keys.py)) and register them (see [here](https://github.com/mir-group/nequip/blob/3db69640fabb684db1e13ec56826c3cb6586fafc/nequip/data/AtomicData.py#L70))

2. Predicting of charges (a per-atom scalar quantity, right?) is possible in Allegro, just that there's no charge equilibration scheme set up to enforce charge conservation. Prediction of dipole moments (per-atom vector quantities, right?) would require some changes in the Allegro module and additional processing modules (to convert spherical tensor features produced by the Allegro module into Cartesian tensor features, such as vectors). It's probably easier to get vectors from `nequip` because you only have to set up the latter processing module since `nequip` already produces tensor features that can be used as-is.
